File: tests/wasm/test_kurtosis.js
/**
 * WASM binding tests for Kurtosis indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32' 
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }
    
    testData = loadTestData();
});

test('Kurtosis partial params', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.kurtosis_js(close, 5);
    assert.strictEqual(result.length, close.length);
});

test('Kurtosis accuracy', async () => {
    
    const hl2 = new Float64Array(testData.hl2);
    
    const result = wasm.kurtosis_js(hl2, 5);
    
    assert.strictEqual(result.length, hl2.length);
    
    
    const expectedLast5 = [
        -0.5438903789933454,
        -1.6848139264816433,
        -1.6331336745945797,
        -0.6130805596586351,
        -0.027802601135927585,
    ];
    
    
    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expectedLast5,
        1e-10,
        "Kurtosis last 5 values mismatch"
    );
    
    
    await compareWithRust('kurtosis', result, 'hl2', { period: 5 });
});

test('Kurtosis default candles', () => {
    
    const hl2 = new Float64Array(testData.hl2);
    
    const result = wasm.kurtosis_js(hl2, 5);
    assert.strictEqual(result.length, hl2.length);
});

test('Kurtosis zero period', () => {
    
    const inputData = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.kurtosis_js(inputData, 0);
    }, /Invalid period/);
});

test('Kurtosis period exceeds length', () => {
    
    const dataSmall = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.kurtosis_js(dataSmall, 10);
    }, /Invalid period/);
});

test('Kurtosis all nan', () => {
    
    const nanData = new Float64Array(20).fill(NaN);
    
    assert.throws(() => {
        wasm.kurtosis_js(nanData, 5);
    }, /All values are NaN/);
});

test('Kurtosis empty input', () => {
    
    const emptyData = new Float64Array([]);
    
    assert.throws(() => {
        wasm.kurtosis_js(emptyData, 5);
    }, /Input data slice is empty/);
});

test('Kurtosis nan prefix', () => {
    
    const nanPrefixData = new Float64Array(30);
    const period = 5;
    nanPrefixData.fill(NaN, 0, 10);
    for (let i = 10; i < 30; i++) {
        nanPrefixData[i] = 50.0 + i * 0.5;
    }
    
    const result = wasm.kurtosis_js(nanPrefixData, period);
    
    assert.strictEqual(result.length, nanPrefixData.length);
    
    
    const expectedNaNCount = 10 + (period - 1);
    for (let i = 0; i < expectedNaNCount; i++) {
        assert(isNaN(result[i]), `Expected NaN at index ${i}`);
    }
    
    
    for (let i = expectedNaNCount; i < result.length; i++) {
        assert(!isNaN(result[i]), `Expected valid value at index ${i}`);
    }
});

test('Kurtosis batch operation', () => {
    
    const hl2 = new Float64Array(testData.hl2);
    
    
    const config = {
        period_range: [5, 5, 0]  
    };
    
    const result = wasm.kurtosis_batch(hl2, config);
    
    assert(result.values);
    assert(result.periods);
    assert.strictEqual(result.rows, 1); 
    assert.strictEqual(result.cols, hl2.length);
    
    
    const expectedLast5 = [
        -0.5438903789933454,
        -1.6848139264816433,
        -1.6331336745945797,
        -0.6130805596586351,
        -0.027802601135927585,
    ];
    
    const last5 = result.values.slice(-5);
    assertArrayClose(
        last5,
        expectedLast5,
        1e-6,
        "Kurtosis batch last 5 values mismatch"
    );
});

test('Kurtosis batch multiple periods', () => {
    
    const close = new Float64Array(testData.close);
    
    const config = {
        period_range: [5, 20, 5]  
    };
    
    const result = wasm.kurtosis_batch(close, config);
    
    assert(result.values);
    assert(result.periods);
    assert.strictEqual(result.rows, 4); 
    assert.strictEqual(result.cols, close.length);
    assert.strictEqual(result.values.length, 4 * close.length);
    
    
    const expectedPeriods = [5, 10, 15, 20];
    assert.deepStrictEqual(result.periods, expectedPeriods);
});

test('Kurtosis fast API', () => {
    
    const close = new Float64Array(testData.close);
    const len = close.length;
    
    
    const inPtr = wasm.kurtosis_alloc(len);
    const outPtr = wasm.kurtosis_alloc(len);
    
    try {
        
        const inputArray = new Float64Array(wasm.__wasm.memory.buffer, inPtr, len);
        inputArray.set(close);
        
        
        wasm.kurtosis_into(inPtr, outPtr, len, 5);
        
        
        const output = new Float64Array(wasm.__wasm.memory.buffer, outPtr, len);
        
        
        const expected = wasm.kurtosis_js(close, 5);
        assertArrayClose(output, expected, 1e-10, "Fast API mismatch");
        
    } finally {
        
        wasm.kurtosis_free(inPtr, len);
        wasm.kurtosis_free(outPtr, len);
    }
});

test('Kurtosis fast API aliasing', () => {
    
    const close = new Float64Array(testData.close);
    const len = close.length;
    
    
    const ptr = wasm.kurtosis_alloc(len);
    
    try {
        
        const dataArray = new Float64Array(wasm.__wasm.memory.buffer, ptr, len);
        dataArray.set(close);
        
        
        wasm.kurtosis_into(ptr, ptr, len, 5);
        
        
        const resultArray = new Float64Array(wasm.__wasm.memory.buffer, ptr, len);
        
        
        const expected = wasm.kurtosis_js(close, 5);
        assertArrayClose(resultArray, expected, 1e-10, "Fast API aliasing mismatch");
        
    } finally {
        
        wasm.kurtosis_free(ptr, len);
    }
});


test('Kurtosis very small dataset', () => {
    
    const smallData = new Float64Array([5.0, 10.0, 15.0, 20.0, 25.0]);
    const period = 5;
    
    const result = wasm.kurtosis_js(smallData, period);
    assert.strictEqual(result.length, smallData.length);
    
    
    for (let i = 0; i < period - 1; i++) {
        assert(isNaN(result[i]), `Expected NaN at index ${i}`);
    }
    
    
    assert(!isNaN(result[period - 1]), `Expected valid value at index ${period - 1}`);
});

test('Kurtosis batch metadata from result', () => {
    
    const close = new Float64Array(testData.close);
    
    const config = {
        period_range: [5, 10, 1]  
    };
    
    const result = wasm.kurtosis_batch(close, config);
    
    assert(result.periods);
    assert.strictEqual(result.periods.length, 6);
    
    
    const expectedPeriods = [5, 6, 7, 8, 9, 10];
    assert.deepStrictEqual(result.periods, expectedPeriods);
});

test('Kurtosis batch edge cases', () => {
    
    const close = new Float64Array(testData.close);
    
    
    const config1 = {
        period_range: [5, 5, 0]
    };
    const result1 = wasm.kurtosis_batch(close, config1);
    assert.strictEqual(result1.rows, 1, "Single parameter should give 1 row");
    
    
    if (close.length > 50) {
        const config2 = {
            period_range: [5, 50, 45]
        };
        const result2 = wasm.kurtosis_batch(close.slice(0, 100), config2);
        assert.strictEqual(result2.rows, 2, "Large step should give 2 rows");
        assert.deepStrictEqual(result2.periods, [5, 50]);
    }
});

test('Kurtosis batch full parameter sweep', () => {
    
    const close = new Float64Array(testData.close.slice(0, 50)); 
    
    
    const config = {
        period_range: [5, 9, 2]
    };
    
    const result = wasm.kurtosis_batch(close, config);
    
    assert(result.values);
    assert(result.periods);
    assert.strictEqual(result.rows, 3); 
    assert.strictEqual(result.cols, 50);
    
    
    const periods = [5, 7, 9];
    for (let i = 0; i < periods.length; i++) {
        const rowStart = i * 50;
        const rowEnd = rowStart + 50;
        const rowData = result.values.slice(rowStart, rowEnd);
        
        const singleResult = wasm.kurtosis_js(close, periods[i]);
        assertArrayClose(
            rowData,
            singleResult,
            1e-10,
            `Period ${periods[i]} mismatch`
        );
    }
});

File: tests/python/test_kurtosis.py
"""
Python binding tests for Kurtosis indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:
    # If not in virtual environment, try to import from installed location
    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close
from rust_comparison import compare_with_rust


class TestKurtosis:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()
    
    def test_kurtosis_partial_params(self, test_data):
        """Test Kurtosis with partial parameters - mirrors check_kurtosis_partial_params"""
        close = test_data['close']
        
        # Test with default period (5)
        result = ta_indicators.kurtosis(close, 5)
        assert len(result) == len(close)
    
    def test_kurtosis_accuracy(self, test_data):
        """Test Kurtosis matches expected values from Rust tests - mirrors check_kurtosis_accuracy"""
        hl2 = test_data['hl2']
        
        result = ta_indicators.kurtosis(hl2, period=5)
        
        assert len(result) == len(hl2)
        
        # Expected values from Rust test
        expected_last_five = [
            -0.5438903789933454,
            -1.6848139264816433,
            -1.6331336745945797,
            -0.6130805596586351,
            -0.027802601135927585,
        ]
        
        # Check last 5 values match expected
        assert_close(
            result[-5:], 
            expected_last_five,
            rtol=1e-6,
            msg="Kurtosis last 5 values mismatch"
        )
        
        # Compare full output with Rust
        compare_with_rust('kurtosis', result, 'hl2', {'period': 5})
    
    def test_kurtosis_default_candles(self, test_data):
        """Test Kurtosis with default parameters - mirrors check_kurtosis_default_candles"""
        hl2 = test_data['hl2']
        
        # Default period is 5
        result = ta_indicators.kurtosis(hl2, 5)
        assert len(result) == len(hl2)
    
    def test_kurtosis_zero_period(self):
        """Test Kurtosis fails with zero period - mirrors check_kurtosis_zero_period"""
        input_data = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.kurtosis(input_data, period=0)
    
    def test_kurtosis_period_exceeds_length(self):
        """Test Kurtosis fails when period exceeds data length - mirrors check_kurtosis_period_exceeds_length"""
        data_small = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.kurtosis(data_small, period=10)
    
    def test_kurtosis_very_small_dataset(self):
        """Test Kurtosis fails with insufficient data - mirrors check_kurtosis_very_small_dataset"""
        single_point = np.array([42.0])
        
        with pytest.raises(ValueError, match="Invalid period|Not enough valid data"):
            ta_indicators.kurtosis(single_point, period=5)
    
    def test_kurtosis_empty_input(self):
        """Test Kurtosis fails with empty input - matches ALMA test pattern"""
        empty = np.array([])
        
        with pytest.raises(ValueError, match="Input data slice is empty"):
            ta_indicators.kurtosis(empty, period=5)
    
    def test_kurtosis_all_nan_input(self):
        """Test Kurtosis fails with all NaN values - matches ALMA test pattern"""
        all_nan = np.full(100, np.nan)
        
        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.kurtosis(all_nan, period=5)
    
    
    def test_kurtosis_nan_handling(self, test_data):
        """Test Kurtosis NaN handling - mirrors check_kurtosis_nan_handling"""
        close = test_data['close']
        period = 5
        
        result = ta_indicators.kurtosis(close, period=period)
        assert len(result) == len(close)
        
        # First (period-1) values should be NaN for warmup
        assert np.all(np.isnan(result[:period-1])), f"Expected NaN in first {period-1} warmup values"
        
        # After warmup period (first 20 values), should not have NaN
        if len(result) > 20:
            non_nan_count = np.count_nonzero(~np.isnan(result[20:]))
            assert non_nan_count == len(result[20:]), "Found unexpected NaN values after warmup"
    
    def test_kurtosis_streaming(self, test_data):
        """Test Kurtosis streaming interface - mirrors check_kurtosis_streaming"""
        close = test_data['close']
        period = 5
        
        # Batch calculation
        batch_result = ta_indicators.kurtosis(close, period=period)
        
        # Streaming calculation
        stream = ta_indicators.KurtosisStream(period=period)
        stream_result = []
        
        for price in close:
            value = stream.update(price)
            if value is None:
                stream_result.append(float('nan'))
            else:
                stream_result.append(value)
        
        # Compare results
        assert len(batch_result) == len(stream_result)
        
        for i, (b, s) in enumerate(zip(batch_result, stream_result)):
            if np.isnan(b) and np.isnan(s):
                continue
            assert_close(
                b, s, 
                rtol=1e-9,
                msg=f"Kurtosis streaming mismatch at index {i}"
            )
    
    def test_kurtosis_batch_default_row(self, test_data):
        """Test Kurtosis batch with default parameters - mirrors check_batch_default_row"""
        hl2 = test_data['hl2']
        
        # Test batch with period range
        result = ta_indicators.kurtosis_batch(hl2, period_range=(5, 5, 0))
        
        assert 'values' in result
        assert 'periods' in result
        
        values = result['values']
        periods = result['periods']
        
        # Should have 1 row (period=5)
        assert values.shape[0] == 1
        assert values.shape[1] == len(hl2)
        assert len(periods) == 1
        assert periods[0] == 5
        
        # Check last 5 values match expected
        expected = [
            -0.5438903789933454,
            -1.6848139264816433,
            -1.6331336745945797,
            -0.6130805596586351,
            -0.027802601135927585,
        ]
        
        assert_close(
            values[0, -5:],
            expected,
            rtol=1e-6,
            msg="Kurtosis batch default row mismatch"
        )
    
    def test_kurtosis_batch_multiple_periods(self, test_data):
        """Test Kurtosis batch with multiple periods"""
        close = test_data['close']
        
        # Test batch with multiple periods
        result = ta_indicators.kurtosis_batch(close, period_range=(5, 20, 5))
        
        assert 'values' in result
        assert 'periods' in result
        
        values = result['values']
        periods = result['periods']
        
        # Should have 4 rows (periods 5, 10, 15, 20)
        assert values.shape[0] == 4
        assert values.shape[1] == len(close)
        assert len(periods) == 4
        assert list(periods) == [5, 10, 15, 20]
        
        # Verify each row has appropriate NaN prefix
        for i, period in enumerate(periods):
            # First (period-1) values should be NaN
            nan_count = np.sum(np.isnan(values[i, :period-1]))
            assert nan_count == period - 1, f"Expected {period-1} NaN values for period {period}"
    
    def test_kurtosis_batch_edge_cases(self, test_data):
        """Test Kurtosis batch with edge case parameters - matches ALMA pattern"""
        close = test_data['close']
        
        # Single parameter (step = 0)
        result1 = ta_indicators.kurtosis_batch(close, period_range=(5, 5, 0))
        assert result1['values'].shape[0] == 1, "Single parameter should give 1 row"
        
        # Large step (only 2 values: 5, 50)
        if len(close) > 50:
            result2 = ta_indicators.kurtosis_batch(close[:100], period_range=(5, 50, 45))
            assert result2['values'].shape[0] == 2, "Large step should give 2 rows"
            assert list(result2['periods']) == [5, 50], "Periods should be [5, 50]"
    
    def test_kurtosis_batch_full_sweep(self, test_data):
        """Test Kurtosis batch full parameter sweep - matches ALMA pattern"""
        close = test_data['close'][:50]  # Use smaller dataset for speed
        
        # Multiple period values: 5, 7, 9
        result = ta_indicators.kurtosis_batch(close, period_range=(5, 9, 2))
        
        assert 'values' in result
        assert 'periods' in result
        
        values = result['values']
        periods = result['periods']
        
        # Should have 3 rows
        assert values.shape[0] == 3
        assert values.shape[1] == 50
        assert list(periods) == [5, 7, 9]
        
        # Verify each row matches individual calculation
        for i, period in enumerate(periods):
            single_result = ta_indicators.kurtosis(close, period=period)
            assert_close(
                values[i, :],
                single_result,
                rtol=1e-10,
                msg=f"Batch row {i} (period={period}) doesn't match single calculation"
            )
    
    def test_kurtosis_kernel_parameter(self, test_data):
        """Test Kurtosis with different kernel parameters"""
        close = test_data['close']
        
        # Test with scalar kernel
        result_scalar = ta_indicators.kurtosis(close, period=5, kernel='scalar')
        assert len(result_scalar) == len(close)
        
        # Test with auto kernel (default)
        result_auto = ta_indicators.kurtosis(close, period=5, kernel='auto')
        assert len(result_auto) == len(close)
        
        # Test invalid kernel
        with pytest.raises(ValueError, match="Unknown kernel"):
            ta_indicators.kurtosis(close, period=5, kernel='invalid')


if __name__ == "__main__":
    pytest.main([__file__, "-v"])